<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>khuongduying - select one slide for visualization</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../Advance/menu3.html" rel="next">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../styles.css">
</head>

<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../GO_WITH_THE_FLOW/2.png" alt="" class="navbar-logo">
    </a>
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">khuongduying</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-tutorials" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Tutorials</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-tutorials">    
        <li>
    <a class="dropdown-item" href="../tutorials.html" rel="" target="">
 <span class="dropdown-text">tutorials.html</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../tutorial-1.html" rel="" target="">
 <span class="dropdown-text">Tutorial 1</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../tutorial-2.html" rel="" target="">
 <span class="dropdown-text">Tutorial 2</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-how-to" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">How-to</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-how-to">    
        <li>
    <a class="dropdown-item" href="../howto.html" rel="" target="">
 <span class="dropdown-text">howto.html</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-fundamentals" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Fundamentals</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-fundamentals">    
        <li>
    <a class="dropdown-item" href="../fundamentals.html" rel="" target="">
 <span class="dropdown-text">fundamentals.html</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-reference" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Reference</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-reference">    
        <li>
    <a class="dropdown-item" href="../reference.html" rel="" target="">
 <span class="dropdown-text">reference.html</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-basic" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Basic</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-basic">    
        <li class="dropdown-header">Basic</li>
        <li>
    <a class="dropdown-item" href="../Basic/menu1.html" rel="" target="">
 <span class="dropdown-text">Basic1</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../Basic/menu2.html" rel="" target="">
 <span class="dropdown-text">Basic2</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-advance" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Advance</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-advance">    
        <li>
    <a class="dropdown-item" href="../Advance/cell2loc_A1_downstream.html" rel="" target="">
 <span class="dropdown-text">select one slide for visualization</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../Advance/menu3.html" rel="" target="">
 <span class="dropdown-text">Advance1</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../Advance/menu4.html" rel="" target="">
 <span class="dropdown-text">Advance2</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="../about.html" rel="" target="">
 <span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../news.html" rel="" target="">
 <span class="menu-text">News</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools ms-auto tools-wide">
    <a href="https://www.linkedin.com/in/khuongduying/" rel="" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-linkedin"></i></a>
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label=""><i class="bi bi-github"></i></a>
      <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="quarto-navigation-tool-dropdown-0">
          <li>
            <a class="dropdown-item quarto-navbar-tools ms-auto-item" href="https://github.com/khuongduying">
            My repositories
            </a>
          </li>
          <li>
            <a class="dropdown-item quarto-navbar-tools ms-auto-item" href="">
            Report a bug
            </a>
          </li>
      </ul>
    </div>
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../Advance/cell2loc_A1_downstream.html">select one slide for visualization</a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto">
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../Advance/cell2loc_A1_downstream.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">select one slide for visualization</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../Advance/menu3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Advance1</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../Advance/menu4.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Advance2</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#identifying-discrete-tissue-regions-by-leiden-clustering" id="toc-identifying-discrete-tissue-regions-by-leiden-clustering" class="nav-link active" data-scroll-target="#identifying-discrete-tissue-regions-by-leiden-clustering">Identifying discrete tissue regions by Leiden clustering</a></li>
  <li><a href="#subset-regions-of-specific-cell-type" id="toc-subset-regions-of-specific-cell-type" class="nav-link" data-scroll-target="#subset-regions-of-specific-cell-type">Subset regions of specific cell-type</a></li>
  <li><a href="#visualizing-marker-genes" id="toc-visualizing-marker-genes" class="nav-link" data-scroll-target="#visualizing-marker-genes">Visualizing marker genes</a>
  <ul class="collapse">
  <li><a href="#cics-marker" id="toc-cics-marker" class="nav-link" data-scroll-target="#cics-marker">CICs marker</a></li>
  </ul></li>
  <li><a href="#colocation" id="toc-colocation" class="nav-link" data-scroll-target="#colocation">Colocation</a>
  <ul class="collapse">
  <li><a href="#estimate-cell-type-specific-expression-of-every-gene-in-the-spatial-data-needed-for-ncem" id="toc-estimate-cell-type-specific-expression-of-every-gene-in-the-spatial-data-needed-for-ncem" class="nav-link" data-scroll-target="#estimate-cell-type-specific-expression-of-every-gene-in-the-spatial-data-needed-for-ncem">Estimate cell-type specific expression of every gene in the spatial data (needed for NCEM)</a></li>
  </ul></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.dev/khuongduying/blob/main/Advance/cell2loc_A1_downstream.ipynb" class="toc-action">Edit this page</a></p><p><a href="https://github.com/khuongduying/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">select one slide for visualization</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<div class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> sys</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> squidpy <span class="im">as</span> sq</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> anndata</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>data_type <span class="op">=</span> <span class="st">'float32'</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> cell2location <span class="im">as</span> c2l</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib <span class="im">as</span> mpl</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> matplotlib <span class="im">import</span> rcParams</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>sc.settings.verbosity <span class="op">=</span> <span class="dv">3</span></span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a>sc.settings.set_figure_params(dpi<span class="op">=</span><span class="dv">80</span>, facecolor<span class="op">=</span><span class="st">"white"</span>)</span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a><span class="co"># silence scanpy that prints a lot of warnings</span></span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> warnings</span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a>warnings.filterwarnings(<span class="st">'ignore'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>/opt/mambaforge/envs/cell2loc_env/lib/python3.9/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
/opt/mambaforge/envs/cell2loc_env/lib/python3.9/site-packages/scvi/_settings.py:63: UserWarning: Since v1.0.0, scvi-tools no longer uses a random seed by default. Run `scvi.settings.seed = 0` to reproduce results from previous versions.
  self.seed = seed
/opt/mambaforge/envs/cell2loc_env/lib/python3.9/site-packages/scvi/_settings.py:70: UserWarning: Setting `dl_pin_memory_gpu_training` is deprecated in v1.0 and will be removed in v1.1. Please pass in `pin_memory` to the data loaders instead.
  self.dl_pin_memory_gpu_training = (</code></pre>
</div>
</div>
<div class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>adata_st <span class="op">=</span> anndata.read_h5ad(</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>            filename<span class="op">=</span><span class="st">"/mnt/rdisk/duydao/PROJECT/SPATIAL/cell2loc_data/adata_st_A1_cell2loc.h5ad"</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>        )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>adata_st</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="3">
<pre><code>AnnData object with n_obs × n_vars = 4961 × 15619
    obs: 'in_tissue', 'array_row', 'array_col', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_hb', 'log1p_total_counts_hb', 'pct_counts_hb', '_indices', '_scvi_batch', '_scvi_labels', 'Alveolar', 'Alveolar Mac', 'CAF', 'CD4+ Treg', 'CD8+ Tem', 'CDKN2A Cancer', 'CXCL1 Cancer', 'Ciliated', 'Endothelial', 'LAMC2 Cancer', 'Lipid-associated Mac', 'Low quality Mac', 'Mast', 'Mature naive B', 'Monocytes', 'NK', 'Naive T', 'Neutrophils', 'Pathological Alveolar', 'Plasma', 'Proliferating Cancer', 'Proliferating Mac', 'Proliferating T/NK', 'SMC', 'SOX2 Cancer', 'cDC2/moDCs', 'pDCs'
    var: 'gene_ids', 'feature_types', 'genome', 'mt', 'hb', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'n_cells', 'feature_name', 'MT_gene'
    uns: '_scvi_manager_uuid', '_scvi_uuid', 'mod', 'spatial'
    obsm: 'MT', 'means_cell_abundance_w_sf', 'q05_cell_abundance_w_sf', 'q95_cell_abundance_w_sf', 'spatial', 'stds_cell_abundance_w_sf'
    layers: 'logcounts'</code></pre>
</div>
</div>
<p>#slide = c2l.utils.select_slide(adata_st, “Visium_FFPE_A1”, batch_key=“patient_region_id”)</p>
<p>with mpl.rc_context({“figure.figsize”: [4.5, 5]}): sc.pl.spatial(adata_st, #slide, cmap=“magma”, color=adata_st.uns[“mod”][“factor_names”], ncols=4, size=1.3, img_key=“hires”, # limit color scale at 99.2% quantile of cell abundance vmin=0, vmax=“p99.2”, )</p>
<section id="identifying-discrete-tissue-regions-by-leiden-clustering" class="level2">
<h2 class="anchored" data-anchor-id="identifying-discrete-tissue-regions-by-leiden-clustering">Identifying discrete tissue regions by Leiden clustering</h2>
<p>The number of KNN neighbours should be adapted to size of dataset and the size of anatomically defined regions</p>
<div class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># compute KNN using the cell2location output stored in adata.obsm</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>sc.pp.neighbors(adata_st, use_rep<span class="op">=</span><span class="st">'q05_cell_abundance_w_sf'</span>,</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>                n_neighbors <span class="op">=</span> <span class="dv">15</span>)</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Cluster spots into regions using scanpy</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>sc.tl.leiden(adata_st, resolution<span class="op">=</span><span class="fl">1.1</span>)</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a><span class="co"># add region as categorical variable</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>adata_st.obs[<span class="st">"region_cluster"</span>] <span class="op">=</span> adata_st.obs[<span class="st">"leiden"</span>].astype(<span class="st">"category"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>computing neighbors
    finished: added to `.uns['neighbors']`
    `.obsp['distances']`, distances for each pair of neighbors
    `.obsp['connectivities']`, weighted adjacency matrix (0:00:03)
running Leiden clustering
    finished: found 15 clusters and added
    'leiden', the cluster labels (adata.obs, categorical) (0:00:01)</code></pre>
</div>
</div>
<p>We can use the location composition similarity graph to build a joint integrated UMAP representation of all section/Visium batches.</p>
<div class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>adata_st.obs[<span class="st">"sample"</span>] <span class="op">=</span> <span class="st">"Visium_FFPE_A1"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Now we use cell2location plotter that allows showing multiple cell types in one panel</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> cell2location.plt <span class="im">import</span> plot_spatial</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="co"># select up to 6 clusters</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>clust_labels <span class="op">=</span> [<span class="st">'CDKN2A Cancer'</span>, <span class="st">'CXCL1 Cancer'</span>, <span class="st">'CD4+ Treg'</span>, <span class="st">'CD8+ Tem'</span>]</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>clust_col <span class="op">=</span> [<span class="st">''</span> <span class="op">+</span> <span class="bu">str</span>(i) <span class="cf">for</span> i <span class="kw">in</span> clust_labels] <span class="co"># in case column names differ from labels</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a><span class="co">#slide = select_slide(adata_vis, 'V1_Human_Lymph_Node')</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> mpl.rc_context({<span class="st">'figure.figsize'</span>: (<span class="dv">15</span>, <span class="dv">15</span>)}):</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>    fig <span class="op">=</span> plot_spatial(</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>        adata<span class="op">=</span>adata_st,</span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>        <span class="co"># labels to show on a plot</span></span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>        color<span class="op">=</span>clust_col, labels<span class="op">=</span>clust_labels,</span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>        show_img<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>        <span class="co"># 'fast' (white background) or 'dark_background'</span></span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>        style<span class="op">=</span><span class="st">'fast'</span>,</span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a>        <span class="co"># limit color scale at 99.2% quantile of cell abundance</span></span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a>        max_color_quantile<span class="op">=</span><span class="fl">0.992</span>,</span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a>        <span class="co"># size of locations (adjust depending on figure size)</span></span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a>        circle_diameter<span class="op">=</span><span class="dv">6</span>,</span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a>        colorbar_position<span class="op">=</span><span class="st">'right'</span></span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true" tabindex="-1"></a>    )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="cell2loc_A1_downstream_files/figure-html/cell-7-output-1.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># compute UMAP using KNN graph based on the cell2location output</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>sc.tl.umap(adata_st, min_dist <span class="op">=</span> <span class="fl">0.3</span>, spread <span class="op">=</span> <span class="dv">1</span>)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="co"># show regions in UMAP coordinates</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> mpl.rc_context({<span class="st">'axes.facecolor'</span>:  <span class="st">'white'</span>,</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>                     <span class="st">'figure.figsize'</span>: [<span class="dv">8</span>, <span class="dv">8</span>]}):</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>    sc.pl.umap(adata_st, color<span class="op">=</span>[<span class="st">'region_cluster'</span>], size<span class="op">=</span><span class="dv">30</span>,</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>               color_map <span class="op">=</span> <span class="st">'RdPu'</span>, ncols <span class="op">=</span> <span class="dv">2</span>, legend_loc<span class="op">=</span><span class="st">'on data'</span>,</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>               legend_fontsize<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>    sc.pl.umap(adata_st, color<span class="op">=</span>[<span class="st">'sample'</span>], size<span class="op">=</span><span class="dv">30</span>,</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>               color_map <span class="op">=</span> <span class="st">'RdPu'</span>, ncols <span class="op">=</span> <span class="dv">2</span>,</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>               legend_fontsize<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a><span class="co"># plot in spatial coordinates</span></span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> mpl.rc_context({<span class="st">'axes.facecolor'</span>:  <span class="st">'black'</span>,</span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a>                     <span class="st">'figure.figsize'</span>: [<span class="dv">8</span>, <span class="dv">8</span>]}):</span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a>    sc.pl.spatial(adata_st, color<span class="op">=</span>[<span class="st">'region_cluster'</span>],</span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a>                  size<span class="op">=</span><span class="fl">1.3</span>, img_key<span class="op">=</span><span class="st">'hires'</span>, alpha<span class="op">=</span><span class="dv">1</span>, bw<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>computing UMAP
    finished: added
    'X_umap', UMAP coordinates (adata.obsm) (0:00:07)</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="cell2loc_A1_downstream_files/figure-html/cell-8-output-2.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="cell2loc_A1_downstream_files/figure-html/cell-8-output-3.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="cell2loc_A1_downstream_files/figure-html/cell-8-output-4.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="subset-regions-of-specific-cell-type" class="level2">
<h2 class="anchored" data-anchor-id="subset-regions-of-specific-cell-type">Subset regions of specific cell-type</h2>
<div class="cell" data-execution_count="34">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>region_cluster_9_11_12 <span class="op">=</span> adata_st[(adata_st.obs[<span class="st">'region_cluster'</span>] <span class="op">==</span> <span class="st">"9"</span>) <span class="op">|</span> </span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>                            (adata_st.obs[<span class="st">'region_cluster'</span>] <span class="op">==</span> <span class="st">"11"</span>) <span class="op">|</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>                            (adata_st.obs[<span class="st">'region_cluster'</span>] <span class="op">==</span> <span class="st">"12"</span>)</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>                            ] </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="35">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot in spatial coordinates</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> mpl.rc_context({<span class="st">'axes.facecolor'</span>:  <span class="st">'black'</span>,</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>                     <span class="st">'figure.figsize'</span>: [<span class="dv">8</span>, <span class="dv">8</span>]}):</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>    sc.pl.spatial(region_cluster_9_11_12, color<span class="op">=</span>[<span class="st">'region_cluster'</span>],</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>                  size<span class="op">=</span><span class="fl">1.3</span>, img_key<span class="op">=</span><span class="st">'hires'</span>, alpha<span class="op">=</span><span class="dv">1</span>, bw<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>                  crop_coord<span class="op">=</span>[<span class="dv">300</span>, <span class="dv">1500</span>, <span class="dv">1000</span>, <span class="dv">3000</span>]</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>                  <span class="co">#crop_coord=[1500, 500, 1500, 500]</span></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>                  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="cell2loc_A1_downstream_files/figure-html/cell-10-output-1.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="35">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot in spatial coordinates</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> mpl.rc_context({<span class="st">'axes.facecolor'</span>:  <span class="st">'black'</span>,</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>                     <span class="st">'figure.figsize'</span>: [<span class="dv">8</span>, <span class="dv">8</span>]}):</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>    sc.pl.spatial(region_cluster_9, color<span class="op">=</span>[<span class="st">'region_cluster'</span>],</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>                  size<span class="op">=</span><span class="fl">1.3</span>, img_key<span class="op">=</span><span class="st">'hires'</span>, alpha<span class="op">=</span><span class="dv">1</span>, bw<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>                  crop_coord<span class="op">=</span>[<span class="dv">300</span>, <span class="dv">1000</span>, <span class="dv">1000</span>, <span class="dv">2400</span>]</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>                  <span class="co">#crop_coord=[1500, 500, 1500, 500]</span></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>                  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="cell2loc_A1_downstream_files/figure-html/cell-11-output-1.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="visualizing-marker-genes" class="level2">
<h2 class="anchored" data-anchor-id="visualizing-marker-genes">Visualizing marker genes</h2>
<div class="cell" data-execution_count="26">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> matplotlib <span class="im">import</span> rcParams</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co">#sc.set_figure_params(dpi=80, color_map='viridis')</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="co">#sc.settings.verbosity = 2</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="co">#sc.logging.print_versions()</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="31">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co">## marker genes</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>marker_genes <span class="op">=</span> [<span class="st">"THEMIS"</span>,<span class="st">"CD8A"</span>, <span class="st">"SELL"</span>, <span class="st">"IL7R"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="23">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>adata_st.layers[<span class="st">'rawcounts'</span>] <span class="op">=</span> adata_st.X</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="32">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>sc.pl.violin(adata_st, marker_genes, </span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>            groupby<span class="op">=</span><span class="st">'region_cluster'</span>,</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>            layer<span class="op">=</span><span class="st">'logcounts'</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="cell2loc_A1_downstream_files/figure-html/cell-16-output-1.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="37">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> mpl.rc_context({<span class="st">'axes.facecolor'</span>:  <span class="st">'black'</span>,</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>                     <span class="st">'figure.figsize'</span>: [<span class="dv">10</span>, <span class="dv">10</span>]}):</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>    ax <span class="op">=</span> sc.pl.stacked_violin(adata_st, marker_genes, </span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>                              groupby<span class="op">=</span><span class="st">'region_cluster'</span>,</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>                              layer<span class="op">=</span><span class="st">'logcounts'</span></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>                              <span class="co">#log=True,</span></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>                            <span class="co">#color=['region_cluster']</span></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>                            <span class="co">#var_group_positions=9, var_group_labels=['CD8+ Tem']</span></span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>                            )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="cell2loc_A1_downstream_files/figure-html/cell-17-output-1.png" class="img-fluid"></p>
</div>
</div>
<section id="cancer-markers" class="level4">
<h4 class="anchored" data-anchor-id="cancer-markers">Cancer markers</h4>
<div class="cell" data-execution_count="38">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co">## marker genes</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>cancer_marker_genes <span class="op">=</span> [<span class="st">"KRT7"</span>,<span class="st">"TTF1"</span>, <span class="st">"KRT16"</span>, <span class="st">"KRT17"</span>, <span class="st">"IL12B"</span>, <span class="st">"KRT5"</span>, <span class="st">"TP63"</span>, <span class="st">"NAPSA"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="41">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> mpl.rc_context({<span class="st">'axes.facecolor'</span>:  <span class="st">'black'</span>,</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>                     <span class="st">'figure.figsize'</span>: [<span class="dv">10</span>, <span class="dv">10</span>]}):</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>    ax <span class="op">=</span> sc.pl.stacked_violin(adata_st, cancer_marker_genes, </span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>                              groupby<span class="op">=</span><span class="st">'region_cluster'</span>,</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>                              layer<span class="op">=</span><span class="st">'logcounts'</span></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>                              <span class="co">#log=True,</span></span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>                            <span class="co">#color=['region_cluster']</span></span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>                            <span class="co">#var_group_positions=9, var_group_labels=['CD8+ Tem']</span></span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a>                            )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="cell2loc_A1_downstream_files/figure-html/cell-19-output-1.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="43">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>region_cluster_1_3_8_12 <span class="op">=</span> adata_st[(adata_st.obs[<span class="st">'region_cluster'</span>] <span class="op">==</span> <span class="st">"1"</span>) <span class="op">|</span> </span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>                            (adata_st.obs[<span class="st">'region_cluster'</span>] <span class="op">==</span> <span class="st">"3"</span>) <span class="op">|</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>                            (adata_st.obs[<span class="st">'region_cluster'</span>] <span class="op">==</span> <span class="st">"8"</span>) <span class="op">|</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>                            (adata_st.obs[<span class="st">'region_cluster'</span>] <span class="op">==</span> <span class="st">"12"</span>)</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>                            ] </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="44">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot in spatial coordinates</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> mpl.rc_context({<span class="st">'axes.facecolor'</span>:  <span class="st">'black'</span>,</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>                     <span class="st">'figure.figsize'</span>: [<span class="dv">8</span>, <span class="dv">8</span>]}):</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>    sc.pl.spatial(region_cluster_1_3_8_12, color<span class="op">=</span>[<span class="st">'region_cluster'</span>],</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>                  size<span class="op">=</span><span class="fl">1.3</span>, img_key<span class="op">=</span><span class="st">'hires'</span>, alpha<span class="op">=</span><span class="dv">1</span>, bw<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>                  <span class="co">#crop_coord=[300, 1500, 1000, 3000]</span></span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>                  <span class="co">#crop_coord=[1500, 500, 1500, 500]</span></span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>                  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="cell2loc_A1_downstream_files/figure-html/cell-21-output-1.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="cics-marker" class="level3">
<h3 class="anchored" data-anchor-id="cics-marker">CICs marker</h3>
<div class="cell" data-execution_count="55">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>cic_marker_genes_1 <span class="op">=</span> [<span class="st">"NET1"</span>, <span class="st">"ALDH3A1"</span>, <span class="st">"PLAT"</span>, <span class="st">"GALNT5"</span>] </span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>cic_marker_genes_2 <span class="op">=</span> [<span class="st">"FAM3C"</span>, <span class="st">"DHRS2"</span> , <span class="st">"AQP3"</span>, <span class="st">"KRT23"</span> , <span class="st">"SYT8"</span>] <span class="co"># , "HLA-B"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="52">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> mpl.rc_context({<span class="st">'axes.facecolor'</span>:  <span class="st">'black'</span>,</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>                     <span class="st">'figure.figsize'</span>: [<span class="dv">10</span>, <span class="dv">10</span>]}):</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>    ax <span class="op">=</span> sc.pl.stacked_violin(adata_st, cic_marker_genes, </span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>                              groupby<span class="op">=</span><span class="st">'region_cluster'</span>,</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>                              layer<span class="op">=</span><span class="st">'logcounts'</span></span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>                              <span class="co">#log=True,</span></span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>                            <span class="co">#color=['region_cluster']</span></span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a>                            <span class="co">#var_group_positions=9, var_group_labels=['CD8+ Tem']</span></span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a>                            )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="cell2loc_A1_downstream_files/figure-html/cell-23-output-1.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="57">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>sc.pl.violin(adata_st, cic_marker_genes_2, </span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>            groupby<span class="op">=</span><span class="st">'region_cluster'</span>,</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>            layer<span class="op">=</span><span class="st">'logcounts'</span></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="cell2loc_A1_downstream_files/figure-html/cell-24-output-1.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="58">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>region_cluster_8_9_11_12 <span class="op">=</span> adata_st[(adata_st.obs[<span class="st">'region_cluster'</span>] <span class="op">==</span> <span class="st">"8"</span>) <span class="op">|</span> </span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>                            (adata_st.obs[<span class="st">'region_cluster'</span>] <span class="op">==</span> <span class="st">"9"</span>) <span class="op">|</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>                            (adata_st.obs[<span class="st">'region_cluster'</span>] <span class="op">==</span> <span class="st">"11"</span>) <span class="op">|</span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>                            (adata_st.obs[<span class="st">'region_cluster'</span>] <span class="op">==</span> <span class="st">"12"</span>)</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>                            ] </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="59">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot in spatial coordinates</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> mpl.rc_context({<span class="st">'axes.facecolor'</span>:  <span class="st">'black'</span>,</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>                     <span class="st">'figure.figsize'</span>: [<span class="dv">8</span>, <span class="dv">8</span>]}):</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>    sc.pl.spatial(region_cluster_8_9_11_12, color<span class="op">=</span>[<span class="st">'region_cluster'</span>],</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>                  size<span class="op">=</span><span class="fl">1.3</span>, img_key<span class="op">=</span><span class="st">'hires'</span>, alpha<span class="op">=</span><span class="dv">1</span>, bw<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>                  <span class="co">#crop_coord=[300, 1500, 1000, 3000]</span></span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>                  <span class="co">#crop_coord=[1500, 500, 1500, 500]</span></span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>                  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="cell2loc_A1_downstream_files/figure-html/cell-26-output-1.png" class="img-fluid"></p>
</div>
</div>
</section>
</section>
<section id="colocation" class="level2">
<h2 class="anchored" data-anchor-id="colocation">Colocation</h2>
<div class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>results_folder<span class="op">=</span><span class="st">"/mnt/rdisk/duydao/PROJECT/SPATIAL/cell2loc_data"</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>run_name <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>results_folder<span class="sc">}</span><span class="ss">/cell2location_map'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="56">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> cell2location <span class="im">import</span> run_colocation</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>res_dict, adata_st <span class="op">=</span> run_colocation(</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>    adata_st,</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>    model_name<span class="op">=</span><span class="st">'CoLocatedGroupsSklearnNMF'</span>,</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>    train_args<span class="op">=</span>{</span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>      <span class="st">'n_fact'</span>: np.arange(<span class="dv">5</span>, <span class="dv">30</span>), <span class="co"># IMPORTANT: use a wider range of the number of factors (5-30)</span></span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a>      <span class="st">'sample_name_col'</span>: <span class="st">'sample'</span>, <span class="co"># columns in adata_vis.obs that identifies sample</span></span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a>      <span class="st">'n_restarts'</span>: <span class="dv">3</span> <span class="co"># number of training restarts</span></span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a>    },</span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a>    <span class="co"># the hyperparameters of NMF can be also adjusted:</span></span>
<span id="cb31-11"><a href="#cb31-11" aria-hidden="true" tabindex="-1"></a>    model_kwargs<span class="op">=</span>{<span class="st">'alpha'</span>: <span class="fl">0.01</span>, <span class="st">'init'</span>: <span class="st">'random'</span>, <span class="st">"nmf_kwd_args"</span>: {<span class="st">"tol"</span>: <span class="fl">0.000001</span>}},</span>
<span id="cb31-12"><a href="#cb31-12" aria-hidden="true" tabindex="-1"></a>    export_args<span class="op">=</span>{<span class="st">'path'</span>: <span class="ss">f'</span><span class="sc">{</span>run_name<span class="sc">}</span><span class="ss">/CoLocatedComb/'</span>}</span>
<span id="cb31-13"><a href="#cb31-13" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>### Analysis name: CoLocatedGroupsSklearnNMF_5combinations_4961locations_27factors
WARNING: saving figure to file /mnt/rdisk/duydao/PROJECT/SPATIAL/cell2loc_data/cell2location_map/CoLocatedComb/CoLocatedGroupsSklearnNMF_4961locations_27factors/spatial/showcell_density_mean_n_fact5_sVisium_FFPE_A1_p99.2.pdf
### Analysis name: CoLocatedGroupsSklearnNMF_6combinations_4961locations_27factors
WARNING: saving figure to file /mnt/rdisk/duydao/PROJECT/SPATIAL/cell2loc_data/cell2location_map/CoLocatedComb/CoLocatedGroupsSklearnNMF_4961locations_27factors/spatial/showcell_density_mean_n_fact6_sVisium_FFPE_A1_p99.2.pdf
### Analysis name: CoLocatedGroupsSklearnNMF_7combinations_4961locations_27factors
WARNING: saving figure to file /mnt/rdisk/duydao/PROJECT/SPATIAL/cell2loc_data/cell2location_map/CoLocatedComb/CoLocatedGroupsSklearnNMF_4961locations_27factors/spatial/showcell_density_mean_n_fact7_sVisium_FFPE_A1_p99.2.pdf
### Analysis name: CoLocatedGroupsSklearnNMF_8combinations_4961locations_27factors
WARNING: saving figure to file /mnt/rdisk/duydao/PROJECT/SPATIAL/cell2loc_data/cell2location_map/CoLocatedComb/CoLocatedGroupsSklearnNMF_4961locations_27factors/spatial/showcell_density_mean_n_fact8_sVisium_FFPE_A1_p99.2.pdf
### Analysis name: CoLocatedGroupsSklearnNMF_9combinations_4961locations_27factors
WARNING: saving figure to file /mnt/rdisk/duydao/PROJECT/SPATIAL/cell2loc_data/cell2location_map/CoLocatedComb/CoLocatedGroupsSklearnNMF_4961locations_27factors/spatial/showcell_density_mean_n_fact9_sVisium_FFPE_A1_p99.2.pdf
### Analysis name: CoLocatedGroupsSklearnNMF_10combinations_4961locations_27factors
WARNING: saving figure to file /mnt/rdisk/duydao/PROJECT/SPATIAL/cell2loc_data/cell2location_map/CoLocatedComb/CoLocatedGroupsSklearnNMF_4961locations_27factors/spatial/showcell_density_mean_n_fact10_sVisium_FFPE_A1_p99.2.pdf
### Analysis name: CoLocatedGroupsSklearnNMF_11combinations_4961locations_27factors
WARNING: saving figure to file /mnt/rdisk/duydao/PROJECT/SPATIAL/cell2loc_data/cell2location_map/CoLocatedComb/CoLocatedGroupsSklearnNMF_4961locations_27factors/spatial/showcell_density_mean_n_fact11_sVisium_FFPE_A1_p99.2.pdf
### Analysis name: CoLocatedGroupsSklearnNMF_12combinations_4961locations_27factors
WARNING: saving figure to file /mnt/rdisk/duydao/PROJECT/SPATIAL/cell2loc_data/cell2location_map/CoLocatedComb/CoLocatedGroupsSklearnNMF_4961locations_27factors/spatial/showcell_density_mean_n_fact12_sVisium_FFPE_A1_p99.2.pdf
### Analysis name: CoLocatedGroupsSklearnNMF_13combinations_4961locations_27factors
WARNING: saving figure to file /mnt/rdisk/duydao/PROJECT/SPATIAL/cell2loc_data/cell2location_map/CoLocatedComb/CoLocatedGroupsSklearnNMF_4961locations_27factors/spatial/showcell_density_mean_n_fact13_sVisium_FFPE_A1_p99.2.pdf
### Analysis name: CoLocatedGroupsSklearnNMF_14combinations_4961locations_27factors
WARNING: saving figure to file /mnt/rdisk/duydao/PROJECT/SPATIAL/cell2loc_data/cell2location_map/CoLocatedComb/CoLocatedGroupsSklearnNMF_4961locations_27factors/spatial/showcell_density_mean_n_fact14_sVisium_FFPE_A1_p99.2.pdf
### Analysis name: CoLocatedGroupsSklearnNMF_15combinations_4961locations_27factors
WARNING: saving figure to file /mnt/rdisk/duydao/PROJECT/SPATIAL/cell2loc_data/cell2location_map/CoLocatedComb/CoLocatedGroupsSklearnNMF_4961locations_27factors/spatial/showcell_density_mean_n_fact15_sVisium_FFPE_A1_p99.2.pdf
### Analysis name: CoLocatedGroupsSklearnNMF_16combinations_4961locations_27factors
WARNING: saving figure to file /mnt/rdisk/duydao/PROJECT/SPATIAL/cell2loc_data/cell2location_map/CoLocatedComb/CoLocatedGroupsSklearnNMF_4961locations_27factors/spatial/showcell_density_mean_n_fact16_sVisium_FFPE_A1_p99.2.pdf
### Analysis name: CoLocatedGroupsSklearnNMF_17combinations_4961locations_27factors
WARNING: saving figure to file /mnt/rdisk/duydao/PROJECT/SPATIAL/cell2loc_data/cell2location_map/CoLocatedComb/CoLocatedGroupsSklearnNMF_4961locations_27factors/spatial/showcell_density_mean_n_fact17_sVisium_FFPE_A1_p99.2.pdf
### Analysis name: CoLocatedGroupsSklearnNMF_18combinations_4961locations_27factors
WARNING: saving figure to file /mnt/rdisk/duydao/PROJECT/SPATIAL/cell2loc_data/cell2location_map/CoLocatedComb/CoLocatedGroupsSklearnNMF_4961locations_27factors/spatial/showcell_density_mean_n_fact18_sVisium_FFPE_A1_p99.2.pdf
### Analysis name: CoLocatedGroupsSklearnNMF_19combinations_4961locations_27factors
WARNING: saving figure to file /mnt/rdisk/duydao/PROJECT/SPATIAL/cell2loc_data/cell2location_map/CoLocatedComb/CoLocatedGroupsSklearnNMF_4961locations_27factors/spatial/showcell_density_mean_n_fact19_sVisium_FFPE_A1_p99.2.pdf
### Analysis name: CoLocatedGroupsSklearnNMF_20combinations_4961locations_27factors
WARNING: saving figure to file /mnt/rdisk/duydao/PROJECT/SPATIAL/cell2loc_data/cell2location_map/CoLocatedComb/CoLocatedGroupsSklearnNMF_4961locations_27factors/spatial/showcell_density_mean_n_fact20_sVisium_FFPE_A1_p99.2.pdf
### Analysis name: CoLocatedGroupsSklearnNMF_21combinations_4961locations_27factors
WARNING: saving figure to file /mnt/rdisk/duydao/PROJECT/SPATIAL/cell2loc_data/cell2location_map/CoLocatedComb/CoLocatedGroupsSklearnNMF_4961locations_27factors/spatial/showcell_density_mean_n_fact21_sVisium_FFPE_A1_p99.2.pdf
### Analysis name: CoLocatedGroupsSklearnNMF_22combinations_4961locations_27factors
WARNING: saving figure to file /mnt/rdisk/duydao/PROJECT/SPATIAL/cell2loc_data/cell2location_map/CoLocatedComb/CoLocatedGroupsSklearnNMF_4961locations_27factors/spatial/showcell_density_mean_n_fact22_sVisium_FFPE_A1_p99.2.pdf
### Analysis name: CoLocatedGroupsSklearnNMF_23combinations_4961locations_27factors
WARNING: saving figure to file /mnt/rdisk/duydao/PROJECT/SPATIAL/cell2loc_data/cell2location_map/CoLocatedComb/CoLocatedGroupsSklearnNMF_4961locations_27factors/spatial/showcell_density_mean_n_fact23_sVisium_FFPE_A1_p99.2.pdf
### Analysis name: CoLocatedGroupsSklearnNMF_24combinations_4961locations_27factors
WARNING: saving figure to file /mnt/rdisk/duydao/PROJECT/SPATIAL/cell2loc_data/cell2location_map/CoLocatedComb/CoLocatedGroupsSklearnNMF_4961locations_27factors/spatial/showcell_density_mean_n_fact24_sVisium_FFPE_A1_p99.2.pdf
### Analysis name: CoLocatedGroupsSklearnNMF_25combinations_4961locations_27factors
WARNING: saving figure to file /mnt/rdisk/duydao/PROJECT/SPATIAL/cell2loc_data/cell2location_map/CoLocatedComb/CoLocatedGroupsSklearnNMF_4961locations_27factors/spatial/showcell_density_mean_n_fact25_sVisium_FFPE_A1_p99.2.pdf
### Analysis name: CoLocatedGroupsSklearnNMF_26combinations_4961locations_27factors
WARNING: saving figure to file /mnt/rdisk/duydao/PROJECT/SPATIAL/cell2loc_data/cell2location_map/CoLocatedComb/CoLocatedGroupsSklearnNMF_4961locations_27factors/spatial/showcell_density_mean_n_fact26_sVisium_FFPE_A1_p99.2.pdf
### Analysis name: CoLocatedGroupsSklearnNMF_27combinations_4961locations_27factors
WARNING: saving figure to file /mnt/rdisk/duydao/PROJECT/SPATIAL/cell2loc_data/cell2location_map/CoLocatedComb/CoLocatedGroupsSklearnNMF_4961locations_27factors/spatial/showcell_density_mean_n_fact27_sVisium_FFPE_A1_p99.2.pdf
### Analysis name: CoLocatedGroupsSklearnNMF_28combinations_4961locations_27factors
WARNING: saving figure to file /mnt/rdisk/duydao/PROJECT/SPATIAL/cell2loc_data/cell2location_map/CoLocatedComb/CoLocatedGroupsSklearnNMF_4961locations_27factors/spatial/showcell_density_mean_n_fact28_sVisium_FFPE_A1_p99.2.pdf
### Analysis name: CoLocatedGroupsSklearnNMF_29combinations_4961locations_27factors
WARNING: saving figure to file /mnt/rdisk/duydao/PROJECT/SPATIAL/cell2loc_data/cell2location_map/CoLocatedComb/CoLocatedGroupsSklearnNMF_4961locations_27factors/spatial/showcell_density_mean_n_fact29_sVisium_FFPE_A1_p99.2.pdf</code></pre>
</div>
</div>
<div class="cell" data-execution_count="42">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Here we plot the NMF weights (Same as saved to `cell_type_fractions_heatmap`)</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>res_dict[<span class="st">'n_fact5'</span>][<span class="st">'mod'</span>].plot_cell_type_loadings()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<pre><code>NameError: name 'res_dict' is not defined</code></pre>
</div>
</div>
<section id="estimate-cell-type-specific-expression-of-every-gene-in-the-spatial-data-needed-for-ncem" class="level3">
<h3 class="anchored" data-anchor-id="estimate-cell-type-specific-expression-of-every-gene-in-the-spatial-data-needed-for-ncem">Estimate cell-type specific expression of every gene in the spatial data (needed for NCEM)</h3>
<section id="load-model" class="level4">
<h4 class="anchored" data-anchor-id="load-model">Load model</h4>
<div class="cell" data-execution_count="41">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>mod <span class="op">=</span> c2l.models.RegressionModel.load(<span class="ss">f"/mnt/rdisk/duydao/PROJECT/SPATIAL/cell2loc_data"</span>, adata_st)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>INFO     File /mnt/rdisk/duydao/PROJECT/SPATIAL/cell2loc_data/model.pt already downloaded                          </code></pre>
</div>
<div class="cell-output cell-output-error">
<pre><code>ValueError: counts is not a valid key in adata.layers.</code></pre>
</div>
</div>
<div class="cell" data-execution_count="25">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>adata_st</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="25">
<pre><code>AnnData object with n_obs × n_vars = 4961 × 15619
    obs: 'in_tissue', 'array_row', 'array_col', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_hb', 'log1p_total_counts_hb', 'pct_counts_hb', '_indices', '_scvi_batch', '_scvi_labels', 'Alveolar', 'Alveolar Mac', 'CAF', 'CD4+ Treg', 'CD8+ Tem', 'CDKN2A Cancer', 'CXCL1 Cancer', 'Ciliated', 'Endothelial', 'LAMC2 Cancer', 'Lipid-associated Mac', 'Low quality Mac', 'Mast', 'Mature naive B', 'Monocytes', 'NK', 'Naive T', 'Neutrophils', 'Pathological Alveolar', 'Plasma', 'Proliferating Cancer', 'Proliferating Mac', 'Proliferating T/NK', 'SMC', 'SOX2 Cancer', 'cDC2/moDCs', 'pDCs', 'leiden', 'region_cluster', 'sample'
    var: 'gene_ids', 'feature_types', 'genome', 'mt', 'hb', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'n_cells', 'feature_name', 'MT_gene'
    uns: '_scvi_manager_uuid', '_scvi_uuid', 'mod', 'spatial', 'neighbors', 'leiden', 'umap', 'region_cluster_colors', 'sample_colors'
    obsm: 'MT', 'means_cell_abundance_w_sf', 'q05_cell_abundance_w_sf', 'q95_cell_abundance_w_sf', 'spatial', 'stds_cell_abundance_w_sf', 'X_umap'
    layers: 'logcounts'
    obsp: 'distances', 'connectivities'</code></pre>
</div>
</div>
<div class="cell" data-execution_count="94">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> plot_genes_per_cell_type(slide, genes, ctypes):</span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a>    n_genes <span class="op">=</span> <span class="bu">len</span>(genes)</span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a>    n_ctypes <span class="op">=</span> <span class="bu">len</span>(ctypes)</span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a>    fig, axs <span class="op">=</span> plt.subplots(</span>
<span id="cb40-10"><a href="#cb40-10" aria-hidden="true" tabindex="-1"></a>        nrows<span class="op">=</span>n_genes, ncols<span class="op">=</span>n_ctypes <span class="op">+</span> <span class="dv">1</span>, figsize<span class="op">=</span>(<span class="fl">4.5</span> <span class="op">*</span> (n_ctypes <span class="op">+</span> <span class="dv">1</span>) <span class="op">+</span> <span class="dv">2</span>, <span class="dv">5</span> <span class="op">*</span> n_genes <span class="op">+</span> <span class="dv">1</span>), squeeze<span class="op">=</span><span class="va">False</span></span>
<span id="cb40-11"><a href="#cb40-11" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb40-12"><a href="#cb40-12" aria-hidden="true" tabindex="-1"></a>    <span class="co"># axs = axs.reshape((n_genes, n_ctypes+1))</span></span>
<span id="cb40-13"><a href="#cb40-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-14"><a href="#cb40-14" aria-hidden="true" tabindex="-1"></a>    <span class="co"># plots of every gene</span></span>
<span id="cb40-15"><a href="#cb40-15" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> j <span class="kw">in</span> <span class="bu">range</span>(n_genes):</span>
<span id="cb40-16"><a href="#cb40-16" aria-hidden="true" tabindex="-1"></a>        <span class="co"># limit color scale at 99.2% quantile of gene expression (computed across cell types)</span></span>
<span id="cb40-17"><a href="#cb40-17" aria-hidden="true" tabindex="-1"></a>        quantile_across_ct <span class="op">=</span> np.array(</span>
<span id="cb40-18"><a href="#cb40-18" aria-hidden="true" tabindex="-1"></a>            [</span>
<span id="cb40-19"><a href="#cb40-19" aria-hidden="true" tabindex="-1"></a>                np.quantile(slide.layers[n][:, slide.var[<span class="st">"feature_name"</span>] <span class="op">==</span> genes[j]].toarray(), <span class="fl">0.992</span>)</span>
<span id="cb40-20"><a href="#cb40-20" aria-hidden="true" tabindex="-1"></a>                <span class="cf">for</span> n <span class="kw">in</span> slide.uns[<span class="st">"mod"</span>][<span class="st">"factor_names"</span>]</span>
<span id="cb40-21"><a href="#cb40-21" aria-hidden="true" tabindex="-1"></a>            ]</span>
<span id="cb40-22"><a href="#cb40-22" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb40-23"><a href="#cb40-23" aria-hidden="true" tabindex="-1"></a>        quantile_across_ct <span class="op">=</span> np.partition(quantile_across_ct.flatten(), <span class="op">-</span><span class="dv">2</span>)[<span class="op">-</span><span class="dv">2</span>]</span>
<span id="cb40-24"><a href="#cb40-24" aria-hidden="true" tabindex="-1"></a>        sc.pl.spatial(</span>
<span id="cb40-25"><a href="#cb40-25" aria-hidden="true" tabindex="-1"></a>            slide,</span>
<span id="cb40-26"><a href="#cb40-26" aria-hidden="true" tabindex="-1"></a>            cmap<span class="op">=</span><span class="st">"magma"</span>,</span>
<span id="cb40-27"><a href="#cb40-27" aria-hidden="true" tabindex="-1"></a>            color<span class="op">=</span>genes[j],</span>
<span id="cb40-28"><a href="#cb40-28" aria-hidden="true" tabindex="-1"></a>            <span class="co"># layer=ctypes[i],</span></span>
<span id="cb40-29"><a href="#cb40-29" aria-hidden="true" tabindex="-1"></a>            gene_symbols<span class="op">=</span><span class="st">"feature_name"</span>,</span>
<span id="cb40-30"><a href="#cb40-30" aria-hidden="true" tabindex="-1"></a>            ncols<span class="op">=</span><span class="dv">4</span>,</span>
<span id="cb40-31"><a href="#cb40-31" aria-hidden="true" tabindex="-1"></a>            size<span class="op">=</span><span class="fl">1.3</span>,</span>
<span id="cb40-32"><a href="#cb40-32" aria-hidden="true" tabindex="-1"></a>            img_key<span class="op">=</span><span class="st">"hires"</span>,</span>
<span id="cb40-33"><a href="#cb40-33" aria-hidden="true" tabindex="-1"></a>            <span class="co"># limit color scale at 99.2% quantile of gene expression</span></span>
<span id="cb40-34"><a href="#cb40-34" aria-hidden="true" tabindex="-1"></a>            vmin<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb40-35"><a href="#cb40-35" aria-hidden="true" tabindex="-1"></a>            vmax<span class="op">=</span><span class="st">"p99.2"</span>,</span>
<span id="cb40-36"><a href="#cb40-36" aria-hidden="true" tabindex="-1"></a>            ax<span class="op">=</span>axs[j, <span class="dv">0</span>],</span>
<span id="cb40-37"><a href="#cb40-37" aria-hidden="true" tabindex="-1"></a>            show<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb40-38"><a href="#cb40-38" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb40-39"><a href="#cb40-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-40"><a href="#cb40-40" aria-hidden="true" tabindex="-1"></a>        <span class="co"># plots of every cell type</span></span>
<span id="cb40-41"><a href="#cb40-41" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(n_ctypes):</span>
<span id="cb40-42"><a href="#cb40-42" aria-hidden="true" tabindex="-1"></a>            sc.pl.spatial(</span>
<span id="cb40-43"><a href="#cb40-43" aria-hidden="true" tabindex="-1"></a>                slide,</span>
<span id="cb40-44"><a href="#cb40-44" aria-hidden="true" tabindex="-1"></a>                cmap<span class="op">=</span><span class="st">"magma"</span>,</span>
<span id="cb40-45"><a href="#cb40-45" aria-hidden="true" tabindex="-1"></a>                color<span class="op">=</span>genes[j],</span>
<span id="cb40-46"><a href="#cb40-46" aria-hidden="true" tabindex="-1"></a>                layer<span class="op">=</span>ctypes[i],</span>
<span id="cb40-47"><a href="#cb40-47" aria-hidden="true" tabindex="-1"></a>                gene_symbols<span class="op">=</span><span class="st">"feature_name"</span>,</span>
<span id="cb40-48"><a href="#cb40-48" aria-hidden="true" tabindex="-1"></a>                ncols<span class="op">=</span><span class="dv">4</span>,</span>
<span id="cb40-49"><a href="#cb40-49" aria-hidden="true" tabindex="-1"></a>                size<span class="op">=</span><span class="fl">1.3</span>,</span>
<span id="cb40-50"><a href="#cb40-50" aria-hidden="true" tabindex="-1"></a>                img_key<span class="op">=</span><span class="st">"hires"</span>,</span>
<span id="cb40-51"><a href="#cb40-51" aria-hidden="true" tabindex="-1"></a>                <span class="co"># limit color scale at 99.2% quantile of gene expression</span></span>
<span id="cb40-52"><a href="#cb40-52" aria-hidden="true" tabindex="-1"></a>                vmin<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb40-53"><a href="#cb40-53" aria-hidden="true" tabindex="-1"></a>                vmax<span class="op">=</span>quantile_across_ct,</span>
<span id="cb40-54"><a href="#cb40-54" aria-hidden="true" tabindex="-1"></a>                ax<span class="op">=</span>axs[j, i <span class="op">+</span> <span class="dv">1</span>],</span>
<span id="cb40-55"><a href="#cb40-55" aria-hidden="true" tabindex="-1"></a>                show<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb40-56"><a href="#cb40-56" aria-hidden="true" tabindex="-1"></a>            )</span>
<span id="cb40-57"><a href="#cb40-57" aria-hidden="true" tabindex="-1"></a>            axs[j, i <span class="op">+</span> <span class="dv">1</span>].set_title(<span class="ss">f"</span><span class="sc">{</span>genes[j]<span class="sc">}</span><span class="ss"> </span><span class="sc">{</span>ctypes[i]<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb40-58"><a href="#cb40-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-59"><a href="#cb40-59" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> fig, axs</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<pre><code>AttributeError: 'dict' object has no attribute 'samples'</code></pre>
</div>
</div>
<div class="cell" data-execution_count="93">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute expected expression per cell type</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>expected_dict <span class="op">=</span> mod.module.model.compute_expected_per_cell_type(</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>    mod.samples[<span class="st">"post_sample_q05"</span>], mod.adata_manager</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Add to anndata layers</span></span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, n <span class="kw">in</span> <span class="bu">enumerate</span>(mod.factor_names_):</span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a>    adata_st.layers[n] <span class="op">=</span> expected_dict[<span class="st">'mu'</span>][i]</span>
<span id="cb42-9"><a href="#cb42-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-10"><a href="#cb42-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Save anndata object with results</span></span>
<span id="cb42-11"><a href="#cb42-11" aria-hidden="true" tabindex="-1"></a>adata_file <span class="op">=</span> <span class="ss">f"</span><span class="sc">{</span>run_name<span class="sc">}</span><span class="ss">/adata_st_est_celltype_specific_expression_A1.h5ad"</span></span>
<span id="cb42-12"><a href="#cb42-12" aria-hidden="true" tabindex="-1"></a>adata_st.write(adata_file)</span>
<span id="cb42-13"><a href="#cb42-13" aria-hidden="true" tabindex="-1"></a>adata_file</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<pre><code>AttributeError: 'dict' object has no attribute 'module'</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="co"># list cell types and genes for plotting</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>ctypes <span class="op">=</span> [<span class="st">"SOX2 Cancer"</span>, <span class="st">"CXCL1 Cancer"</span>, <span class="st">"CD8+ Tem"</span>]</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>genes <span class="op">=</span> [<span class="st">'CD3D'</span>, <span class="st">'THEMIS'</span>]</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> mpl.rc_context({<span class="st">'axes.facecolor'</span>:  <span class="st">'black'</span>}):</span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a>    <span class="co"># select one slide</span></span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>    <span class="co">#slide = select_slide(adata_st, 'V1_Human_Lymph_Node')</span></span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a>    <span class="co">#from tutorial_utils import plot_genes_per_cell_type</span></span>
<span id="cb44-10"><a href="#cb44-10" aria-hidden="true" tabindex="-1"></a>    plot_genes_per_cell_type(adata_st, genes, ctypes)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


</section>
</section>
</section>

<a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
  </div>
  <div class="nav-page nav-page-next">
      <a href="../Advance/menu3.html" class="pagination-link">
        <span class="nav-page-text">Advance1</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">Copyright 2024, Duy Dao</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
    <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/khuongduying">
      <i class="bi bi-github" role="img">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://www.linkedin.com/in/khuongduying/">
      <i class="bi bi-linkedin" role="img">
</i> 
    </a>
  </li>  
</ul>
    </div>
  </div>
</footer>



</body></html>